### PR TITLE
change(ci): Generate mainnet checkpoints in CI

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

In #6367 we want to generate checkpoints in CI. This is the mainnet CI and Rust test harness part of that ticket.

### Complex Code or Requirements

This is just test code.

## Solution

CI workflow:
- [Add Mainnet CI workflow for zebra-checkpoints](https://github.com/ZcashFoundation/zebra/commit/de274a780a8c57b192c8708ae708adaf69d276b8)

Rust test harness:
- [Add Mainnet and Testnet zebra-checkpoints test harnesses](https://github.com/ZcashFoundation/zebra/commit/28f31308cfb6ed2ed81fe04fb2ffad9159c00f55)
- [Build zebra-utils binaries for zebrad integration tests](https://github.com/ZcashFoundation/zebra/pull/6550/commits/58f3f748d1ec92b73e51fb2e2e0749124b8112d5)
- [Start generating checkpoints from the last compiled-in checkpoint](https://github.com/ZcashFoundation/zebra/pull/6550/commits/9f219d32dc0059b4b99429e109175a7581844b9a)
- [Try zebra-checkpoints in the system path first, then the cargo path](https://github.com/ZcashFoundation/zebra/pull/6550/commits/fbe65bfc6b1a1713f20a5ad23d5d6f507674daca)

Docker test harness:
- [Add zebra-checkpoints test entrypoints](https://github.com/ZcashFoundation/zebra/commit/25278467b36d4f11aae5720971db23866bda8b97)
- [Add documentation for how to run these tests individually](https://github.com/ZcashFoundation/zebra/pull/6550/commits/8bb885ef33dad0d8020a400f6a6805b4b0e6ca6c)

zebra-checkpoints binary:
- [Add zebra-checkpoints to test docker images](https://github.com/ZcashFoundation/zebra/commit/a33d0cc0919fc5528e2ebec8ff34648b5b21385f)
- [Enable zebra-checkpoints feature in the test image](https://github.com/ZcashFoundation/zebra/pull/6550/commits/fc07390e723655ccd8490d25247f67a523de3c1a)

Logging:
- [Silence zebrad logs while generating checkpoints](https://github.com/ZcashFoundation/zebra/pull/6550/commits/2319726c5ae176d599c3107185b27f7fd571cdb3)
- [Log a startup message to standard error when launching zebra-checkpoints](https://github.com/ZcashFoundation/zebra/pull/6550/commits/873bf21b72dbf8cad701c4c60b7691fce504c5bf)

Related changes:
- [Use the same features for (almost) all the docker tests](https://github.com/ZcashFoundation/zebra/pull/6550/commits/9cbe956407d0e07063cc5b10d92dc548c45e8278)
- [Log the original test command name when showing stdout and stderr](https://github.com/ZcashFoundation/zebra/pull/6550/commits/d9768a12b741168b173d7912dd7620b47a8acb50)

And a bunch of minor fixes.

## Review

This is a routine CI change.

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
    - [ ] Does it change concurrent code, unsafe code, or consensus rules?
  - [ ] How do you know it works? Does it have tests?

## Follow Up Work

- [ ] Admin: Require the new mainnet job to pass before a PR can merge

Next PR:
- Add a workflow for testnet

### CODE REVIEW

File: docker/Dockerfile
Comment (teor2345) [inline docker/Dockerfile:85]:
```suggestion
# Keep these in sync with:
# https://github.com/ZcashFoundation/zebra/blob/main/.github/workflows/build-docker-image.yml#L42
ARG FEATURES="sentry"
```

File: docker/entrypoint.sh
Comment (teor2345) [inline docker/entrypoint.sh]: TODO: check the script output then delete this

File: zebrad/tests/common/checkpoints.rs
Comment (teor2345) [inline zebrad/tests/common/checkpoints.rs]: TODO: why does it fail here?

File: zebrad/tests/common/checkpoints.rs
Comment (arya2) [inline zebrad/tests/common/checkpoints.rs]:
```suggestion
    // Mainnet and testnet checkpoints always have at least one leading zero in their hash.
```

File: docker/Dockerfile
Comment (arya2) [inline docker/Dockerfile:112]: Could we add 'sentry' in to this features argument in release-binaries.yml:
https://github.com/ZcashFoundation/zebra/blob/8a9113c4e2fc3cd551f5cf94c3b8177591a54eae/.github/workflows/release-binaries.yml#L49

Nitpick:
```suggestion
RUN cargo chef cook --release --features "${ENTRYPOINT_FEATURES}" --workspace --recipe-path recipe.json
```

File: zebrad/tests/acceptance.rs
Comment (arya2) [inline zebrad/tests/acceptance.rs:2217]: Could we add an example of how to run this locally,

```rust
//! ## Zebra checkpoint generation tests
//!
//! Example of how to generate checkpoints with Zebrad's checkpoints tests:
//!
//! ZEBRA_CACHED_STATE_DIR=~/path/to/zebra/chain cargo test checkpoints_mainnet --features zebra-checkpoints --release  -- --ignored --nocapture

```

File: zebrad/tests/common/checkpoints.rs
Comment (teor2345) [reply zebrad/tests/common/checkpoints.rs]: This code is either fixed or deleted

File: docker/Dockerfile
Comment (teor2345) [reply docker/Dockerfile:112]: > Could we add 'sentry' in to this features argument in release-binaries.yml:

Done!

> Nitpick:

It's meant to be separate so it can be overridden separately. I'll add a note.

File: docker/Dockerfile
Comment (teor2345) [inline docker/Dockerfile:89]:
```suggestion
# Use ENTRYPOINT_FEATURES to override the specific features used to run tests in entrypoint.sh,
# separately from the test and production image builds.
ENV ENTRYPOINT_FEATURES "$TEST_FEATURES $FEATURES"
```

File: zebrad/tests/acceptance.rs
Comment (teor2345) [reply zebrad/tests/acceptance.rs:2217]: Happy to do this, but I'd like to use `entrypoint.sh`:
```sh
ENTRYPOINT_FEATURES="zebra-checkpoints" GENERATE_CHECKPOINTS_MAINNET=1 ZEBRA_CACHED_STATE_DIR=/var/cache/zebrad-cache  entrypoint.sh
ENTRYPOINT_FEATURES="zebra-checkpoints" GENERATE_CHECKPOINTS_TESTNET=1 ZEBRA_CACHED_STATE_DIR=/var/cache/zebrad-cache  entrypoint.sh
```

That way we don't have to update the docs as the script changes.

Is it ok if I put it with the rest of the examples?
https://github.com/ZcashFoundation/zebra/blob/8a9113c4e2fc3cd551f5cf94c3b8177591a54eae/zebrad/tests/acceptance.rs#L104-L110

Comment (teor2345) [reply zebrad/tests/acceptance.rs:2217]: Done in [2ca16c0](https://github.com/ZcashFoundation/zebra/pull/6550/commits/2ca16c09baf9118f4039e1fa3356e53af0a50a26)

**Comment by teor2345** [comment]
This works locally for me, using the same `entrypoint.sh` script.

There have been some weird issues where there's no `zebra-checkpoints` process output. I've seen it occasionally locally, and a lot in CI. They are a bit hard to diagnose, because the process doesn't fail. It just runs and exits, without any other output.

I've just pushed a workaround.

It should look something like:
Screenshot 2023-04-21 at 15 47 55

Screenshot 2023-04-21 at 15 48 27

**Comment by teor2345** [comment]
This is really strange - the test harness works locally, but in Docker + Google Cloud, `zebra-checkpoints` seems to immediately close its standard error as soon as it launches. (Or possibly never launch?)

```
Fri, 21 Apr 2023 07:29:09 GMT 2023-04-21T07:29:09.014654Z  INFO spawn_zebra_checkpoints_direct{network=Mainnet test_type=UpdateZebraCachedStateWithRpc zebrad_rpc_address=127.0.0.1:36529 last_checkpoint="2056451"}: Trying to launch zebra-checkpoints from cargo path... zebra_checkpoints_path="/opt/zebrad/target/release/zebra-checkpoints" args=Arguments({"--transport": Some("direct"), "--addr": Some("127.0.0.1:36529"), "--last-checkpoint": Some("2056451")})
Fri, 21 Apr 2023 07:29:09 GMT
Fri, 21 Apr 2023 07:29:09 GMT Child Stderr:
Fri, 21 Apr 2023 07:29:09 GMT [sentry] initialized disabled sentry client due to disabled or invalid DSN
Fri, 21 Apr 2023 07:29:09 GMT
Fri, 21 Apr 2023 07:29:09 GMT Error: 
Fri, 21 Apr 2023 07:29:09 GMT    0: stderr of command did not log any matches for the given regex,
Fri, 21 Apr 2023 07:29:09 GMT       within the 11h command timeout
Fri, 21 Apr 2023 07:29:09 GMT
Fri, 21 Apr 2023 07:29:09 GMT Location:
Fri, 21 Apr 2023 07:29:09 GMT    /opt/zebrad/zebra-test/src/command.rs:923
Fri, 21 Apr 2023 07:29:09 GMT
Fri, 21 Apr 2023 07:29:09 GMT Match Regex:
Fri, 21 Apr 2023 07:29:09 GMT    [
Fri, 21 Apr 2023 07:29:09 GMT        "calculating checkpoints",
Fri, 21 Apr 2023 07:29:09 GMT    ]
Fri, 21 Apr 2023 07:29:09 GMT
Fri, 21 Apr 2023 07:29:09 GMT Command:
Fri, 21 Apr 2023 07:29:09 GMT    "/opt/zebrad/target/release/zebra-checkpoints" "--transport" "direct" "--addr" "127.0.0.1:36529" "--last-checkpoint" "2056451"
```

https://github.com/ZcashFoundation/zebra/actions/runs/4762133164/jobs/8464755724?pr=6550#step:8:218

**Comment by teor2345** [comment]
When I run `zebra-checkpoints --help` locally, it gives me usage info. But when I run it in CI I get no output:

> 2023-04-24T01:28:34.846314Z  INFO spawn_zebra_checkpoints_direct{network=Mainnet test_type=UpdateZebraCachedStateWithRpc zebrad_rpc_address=127.0.0.1:41547 last_checkpoint="2056451"}: Trying to launch zebra-checkpoints by searching system $PATH... zebra_checkpoints_path="zebra-checkpoints" args=Arguments({"--transport": Some("direct"), "--addr": Some("127.0.0.1:41547"), "--last-checkpoint": Some("2056451")}) temp_dir="/tmp/zebrad_testsQtGqof" system_path=Ok("/opt/zebrad/.cargo/bin:/usr/local/cargo/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin") usr_local_zebra_checkpoints_info=Ok(Metadata { file_type: FileType(FileType { mode: 33261 }), is_dir: false, is_file: true, permissions: Permissions(FilePermissions { mode: 33261 }), modified: Ok(SystemTime { tv_sec: [1682299330](tel:1682299330), tv_nsec: 0 }), accessed: Ok(SystemTime { tv_sec: [1682299330](tel:1682299330), tv_nsec: 0 }), created: Ok(SystemTime { tv_sec: [1682299613](tel:1682299613), tv_nsec: 482256533 }), .. })

(usage output should be here)

https://github.com/ZcashFoundation/zebra/actions/runs/4781308585/jobs/8500061799#step:8:179

It seems like there's something wrong with the `zebra-checkpoints` binary, or with the test harness. I'll run the test Docker image locally to find out.

**Comment by teor2345** [comment]
The Docker command used by the failing test was:
```sh
gcloud compute ssh generate-checkpoints-mainnet-6550-merge-5df775d \
  --zone us-central1-f \
  --ssh-flag="-o ServerAliveInterval=5" \
  --ssh-flag="-o ConnectionAttempts=20" \
  --ssh-flag="-o ConnectTimeout=5" \
  --command \
  "\
  sudo docker run \
  --name generate-checkpoints-mainnet \
  --tty \
  --detach \
  -e GENERATE_CHECKPOINTS_MAINNET=1 -e ZEBRA_FORCE_USE_COLOR=1 -e ZEBRA_CACHED_STATE_DIR=/var/cache/zebrad-cache \
  --mount type=volume,src=generate-checkpoints-mainnet-5df775d,dst=/var/cache/zebrad-cache \
  us-docker.pkg.dev/zealous-zebra/zebra/zebrad-test:sha-5df775d \
  "
```

https://github.com/ZcashFoundation/zebra/actions/runs/4781308585/jobs/8500043278#step:8:15

So I ran this command locally to check the Docker image:
```sh
docker run --tty -e GENERATE_CHECKPOINTS_MAINNET=1 -e ZEBRA_FORCE_USE_COLOR=1 -e ZEBRA_CACHED_STATE_DIR=/var/cache/zebrad-cache --volume="$HOME/.cache/zebra":/var/cache/zebrad
-cache us-docker.pkg.dev/zealous-zebra/zebra/zebrad-test:sha-5df775d   
```

And it fails the same way. So the issue is with either the Docker image, or the test harness. It's not in Google Cloud.

I ran an interactive shell in the Docker image:
```sh
docker run --tty -i -e GENERATE_CHECKPOINTS_MAINNET=1 -e ZEBRA_FORCE_USE_COLOR=1 -e ZEBRA_CACHED_STATE_DIR=/var/cache/zebrad-cache --volume="$HOME/.cache/zebra":/var/cache/zebrad
-cache --entrypoint bash us-docker.pkg.dev/zealous-zebra/zebra/zebrad-test:sha-5df775d   
```

`zebra-checkpoints` is 1.9 MB in the Docker image, compared with 36 MB locally. That's almost exactly the same size as `zebra-cli` in the Docker image. (It's 32 bytes less, which is the difference in the length of the executable name times 4).

So it looks like they are both Rust binaries with empty `fn main {}`.

**Comment by teor2345** [comment]
Turns out that the `zebra-checkpoints` binary was from a fake empty `main.rs` file generated by the Docker caching layer. And since the real `main.rs` wasn't being built correctly by the tests, we got the empty file as a binary instead.

 This is something we should watch out for in future tests.

Edit: Here's the `cargo chef` bug:
https://github.com/LukeMathWalker/cargo-chef/issues/212

**Comment by teor2345** [comment]
It works on mainnet! 🎉

https://github.com/ZcashFoundation/zebra/actions/runs/4782664255/jobs/8502843980?pr=6550#step:8:248

IMG_2458

**Review by gustavovalverde — COMMENTED** [review]
LGTM!

I really liked the improved version of the Dockerfile using `${TEST_FEATURES}` and `${FEATURES}`

I'll leave the final approval to @arya2, as there's Rust code to review.

**Review by arya2 — DISMISSED** [review]

**Comment by mpguerra** [comment]
@Mergifyio refresh

**Comment by teor2345** [comment]
@mergifyio refresh

**Comment by teor2345** [comment]
Anyone can re-approve this PR, I just rebased it after PR #6516 merged.

**Review by gustavovalverde — APPROVED** [review]
Re-approving

**Pull request merged**
